### PR TITLE
Add support for listing and downloading external multimedia

### PR DESCRIPTION
**Pull request by bnjmnt4n**
This PR is stacked on #643.

I cleaned up some of the code and rebased on top of #643, however there's still quite a bit of mess. I likely won't have time to come back to this soon, so I thought I would just send a PR in case anyone else is interested in finishing this up.

### CODE REVIEW

File: src/external_multimedia.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                    })
+                    .await?;
+
+                // TODO: NO HARDCODE!>! :P
+                let query = response.url().fragment();
+                let folder_id = if let Some(fragment) = query {
+                    Some(&fragment[12..fragment.len() - 3])
+                } else {
+                    None
+                };
```
Comment (bnjmnt4n) [inline src/external_multimedia.rs:155]: It was taking a little too long for me to figure out how to parse the fragment string, so I just did some hardcoded slices of the fragment string :'(

File: src/external_multimedia.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                            .results
+                            .into_iter()
+                            .map(|m| ExternalMultimediaVideo {
+                                html_url: m.viewer_url,
+                                path: channel_path.join(Self::make_mp4_extension(Path::new(
+                                    &sanitise_filename(&m.session_name),
+                                ))),
+                                // TODO: last updated for external multimedia
+                                // last_updated: parse_time(&m.last_updated_date),
+                                last_updated: SystemTime::UNIX_EPOCH,
```
Comment (bnjmnt4n) [inline src/external_multimedia.rs:192]: There doesn't seem to be a way to get the last updated date in the API, so I hardcoded this to the Unix epoch for now. This means you can't really check for updates (can external multimedia even be updated?)

File: src/external_multimedia.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                None,
+            )
+            .await?;
+
+        match multimedia_resp.data {
+            Some(channels) => future::join_all(
+                channels
+                    .into_iter()
+                    // Only load external multimedia resources
+                    // TODO: figure out how to integrate with `multimedia`
```
Comment (bnjmnt4n) [inline src/external_multimedia.rs:101]: Not sure if this should somehow be integrated into `multimedia.rs`.

**Review by bnjmnt4n — COMMENTED** [review]
This PR works for me in listing and downloading external multimedia, but there is a major issue that happens intermittently in which one of the futures which appears to block completion of the operation. I couldn't really debug it since I'm not familiar with that.

**Review by btzy — COMMENTED** [review]
Thanks for the PR!  Looks good already, though maybe there's a bit of code duplication with multimedia.rs and weblecture.rs.  (I won't be able to work on it this week though, and after that I'll probably lose access to this semester's modules so it'll have to wait until next semester if not done by then.)

Maybe some API endpoint is returning an error intermittently, which would cause `infinite_retry_http()` to be stuck in an infinite loop.  Afaik we needed to retry because the API will reject our request if we make too many requests in a short time (see #635), but the current implementation of `infinite_retry_http()` doesn't distinguish that error from other more generic errors (in which case we should probably not retry, or retry a limited number of times).  Not sure, but it might help to log the returned error messages from the HTTP library.

**Comment by btzy** [comment]
I refactored this PR somewhat and combined multimedia and external multimedia slightly.

I can't reproduce the blocking issue that you mentioned though.  I've tried at least 3 times, and it always worked.  Not sure why you have that issue.

**Comment by bnjmnt4n** [comment]
Thanks for finishing up this PR! Looks like the blocking issue doesn't really occur anymore - maybe there was some bug with my code somewhere.

Another thing that could be handled in the future: seems like external media on Panopto could have the same name, so maybe an ID could be appended to the filename.
```
CS2040S/Multimedia/Lectures/zoom_0.mp4
CS2040S/Multimedia/Lectures/zoom_0.mp4
CS2040S/Multimedia/Lectures/zoom_0.mp4
CS2040S/Multimedia/Lectures/zoom_0.mp4
CS2040S/Multimedia/Lectures/zoom_0.mp4
```

**Comment by btzy** [comment]
Yeah, I've just realised that the same can be the case for Zoom conferences.  Probably should refactor the CLI to take this into account for all kinds of resources.

**Pull request merged**